File: _tbx/lp_tbx/estimateLPz.m
function [LP] = estimateLPz(data, z, h, p, c_case, exdata, alpha)

% Prep parameters/containers
t = size(data,1);
n = size(data,2);
gamma = NaN*ones(h,n);
gammabands = NaN*ones(h,n,2);
if isempty(alpha)
    alpha = 90;
end
zscore = abs(mynorminv((1-alpha/100)/2));


%% Y matrix
Y = data;

%% X matrix: Generate lags
X = [];
for pp = 1:p
    X = [X, [NaN*ones(pp, n); data(1:end-pp,:)] ];
end
% add constant and/or trend to X matrix
if c_case == 0
    c = [];
elseif c_case == 1
    c = ones(size(X,1),1); 
elseif c_case == 2
    c = [ones(size(X,1),1), [1:1:size(X)]' ];
else
    print('c_case variable needs to be set to 0, 1, or 2.')
end
% add exogenous variables
if ~isempty(exdata)
    Xex = [];
    for pp = 1:p
        Xex = [Xex, [NaN*ones(pp, size(exdata,2)); exdata(1:end-pp,:)]  ];
    end
else
    Xex = [];
end
X = [c, X, Xex];


%% Estimation
disp('Progress \n')
counter = 1;
nloops = n*h;
for nn = 1:1:n
    for hh = 0:1:(h-1)

        % prepare LHS variable (changes in each loop while RHS stays
        % constant)
        lhs = [Y(hh+1:end,nn); NaN*ones(hh,1)];

        % drop observations with any NA's
        nona = logical(sum(isnan([lhs,z,X,Xex]), 2) == 0);
        lhs = lhs(nona,:);
        rhs = [z(nona,:), X(nona,:)];
        
        % reduced-form estimation (OLS with Newey-West standard errors)
        reg = nwest(lhs,rhs,p); % regression step
        vcov = rhs'*rhs;
        gamma(hh+1,nn) = reg.beta(1);
        gammabands(hh+1,nn,:) = reg.beta(1) + [-1 1]*zscore*reg.se(1);
        
        % show progress report
        fprintf(1,'\b\b\b\b%3.0f%%',100*(counter/nloops)); 
        counter = counter + 1;

    end
end
fprintf('\n');

%% Housekeeping
LP.data = data;
LP.X = X;
LP.Xex = Xex;
LP.c_case = c_case;
LP.p = p;
LP.t = t;
LP.n = n;
LP.gamma = gamma;
LP.gammabands = gammabands;
LP.h = h;

File: _tbx/lp_tbx/estimateLPz.R
estimateLPz <- function(data, z, h, p, c_case, exdata, alpha=90, NWSE=T){
  
  # Prep parameters/containers
  t <- nrow(data)
  n <- ncol(data)
  gamma <- matrix(NA, nrow=h, ncol=n)
  gammabands <- array(NA, dim=c(h,n,2))

  zscore <- qnorm(1-(1-alpha/100)/2)
  
  # Y matrix
  Y <- unname(as.matrix(data))
  
  # X matrix: Generate lags
  X <- NULL
  for (pp in 1:p){
    X <- cbind(X, rbind(matrix(NA, nrow=pp, ncol=n), Y[1:(nrow(data)-pp),]))
  }
  
  # add constant and/or trend to X matrix
  if (c_case == 0){
    c <- NULL
  } else if (c_case == 1){
    c <- matrix(1, nrow=nrow(X), ncol=1)
  } else if (c_case == 2){
    c <- cbind(matrix(1, nrow=nrow(X), ncol=1), matrix(1:nrow(X), nrow=nrow(X), ncol=1))
  } else {
    print('c_case variable needs to be set to 0, 1 or 2.')
  }
  
  # add exogenous variables
  Xex <- NULL
  if (!is.null(exdata)){
    for (pp in 1:p){
      Xex <- cbind(Xex, rbind(matrix(NA, nrow=pp, ncol=ncol(Xex)), exdata[1:(nrow(data)-pp),]))
    }
    Xex <- Xex[(p+1):nrow(Xex), ]
  }
  X <- cbind(c, X, Xex)
  
  ## Estimation
  print('Progress...')
  counter <- 1
  pb <- txtProgressBar(min = 0, max = n*h, initial = 0, style=3) 
  for (nn in 1:n){
    for (hh in 0:(h-1)){
      
      # prepare LHS variable (changes in each loop while RHS stays constant)
      lhs = as.matrix(c(Y[(hh+1):nrow(Y),nn], rep(NA, hh)), ncol=1)
      
      # drop observations with any NA's
      regdata <- data.frame(cbind(lhs, z, X))
      nona <- rowSums(is.na(regdata)) == 0
      regdata <- regdata[nona, ]
      
      # reduced-form estimation (OLS with Newey-West standard errors)
      reg <- lm(X1 ~ 0 + ., data=regdata)
      if (NWSE){
        vcovNW <- unname(NeweyWest(reg))
        se <- sqrt(diag(vcovNW))
      } else {
        se <- summary(reg)$coefficients[,2]
      }
      
      gamma[hh+1,nn] <- reg$coefficients[1]
      gammabands[hh+1,nn,1] <- reg$coefficients[1] - zscore*se[1]
      gammabands[hh+1,nn,2] <- reg$coefficients[1] + zscore*se[1]
      
      # show progress report
      setTxtProgressBar(pb,counter)
      counter <- counter + 1
      
    }
  }
  
  # Housekeeping
  return(LP = list('data' = data,
                   'exdata' = exdata,
                   'X' = X,
                   'Xex' = Xex,
                   'c_case' = c_case,
                   'p' = p,
                   't' = t,
                   'n' = n,
                   'gamma' = gamma,
                   'gammabands' = gammabands,
                   'h' = h))
  
}